File: ios/Oriveo/Oriveo/Core/Providers/LocalEngineRuntime.swift
import Foundation

nonisolated struct LocalRuntimeSettings: Codable, Equatable, Sendable {
    var contextLength: Int?
    var keepAlive: String?
    var speculativeModel: String?
    var cacheReuseTokens: Int?
}

final class LocalRuntimeSettingsStore: @unchecked Sendable {
    static let shared = LocalRuntimeSettingsStore()
    private let defaults: UserDefaults
    private let key = "local_runtime_settings.v1"
    private let lock = NSLock()

    init(defaults: UserDefaults = .standard) { self.defaults = defaults }

    static func fingerprint(endpoint: String, engine: LocalEngineKind) -> String {
        var hash: UInt64 = 14_695_981_039_346_656_037
        for byte in "\(engine.rawValue)|\(endpoint)".utf8 { hash = (hash ^ UInt64(byte)) &* 1_099_511_628_211 }
        return String(hash, radix: 16)
    }

    func load(endpointFingerprint: String) -> LocalRuntimeSettings {
        lock.lock(); defer { lock.unlock() }
        return recordsLocked()[endpointFingerprint] ?? .init()
    }

    func save(_ settings: LocalRuntimeSettings, endpointFingerprint: String) {
        lock.lock(); defer { lock.unlock() }
        var records = recordsLocked()
        records[endpointFingerprint] = settings
        if let data = try? JSONEncoder().encode(records) { defaults.set(data, forKey: key) }
    }

    func remove(endpointFingerprint: String) {
        lock.lock(); defer { lock.unlock() }
        var records = recordsLocked()
        records.removeValue(forKey: endpointFingerprint)
        if let data = try? JSONEncoder().encode(records) { defaults.set(data, forKey: key) }
    }

    private func recordsLocked() -> [String: LocalRuntimeSettings] {
        guard let data = defaults.data(forKey: key) else { return [:] }
        return (try? JSONDecoder().decode([String: LocalRuntimeSettings].self, from: data)) ?? [:]
    }
}

nonisolated struct LocalRuntimeSnapshot: Equatable, Sendable {
    var health: LocalEngineState?
    var loadedModelIDs: [String]
    var ttftMilliseconds: Double?
    var tokensPerSecond: Double?
    var contextUsed: Int?
    var contextLimit: Int?
    var queueDepth: Int?
    var cpuPercent: Double?
    var gpuPercent: Double?
}

nonisolated enum LocalPromptPreflight: Equatable, Sendable {
    case supported(tokens: Int, contextLimit: Int?, exceedsContext: Bool)
    case unavailable
}

/// Production parsers keep unavailable observations nil. Zero is preserved only when the engine actually reports zero.
enum LocalRuntimeParser {
    static func snapshot(
        engine: LocalEngineKind,
        statusJSON: Any?,
        metricsText: String? = nil,
        timingsJSON: Any? = nil
    ) -> LocalRuntimeSnapshot {
        let object = statusJSON as? [String: Any] ?? [:]
        let slots = object["slots"] as? [[String: Any]] ?? (statusJSON as? [[String: Any]]) ?? []
        let models = object["models"] as? [[String: Any]] ?? object["data"] as? [[String: Any]] ?? []
        let loaded = models.compactMap { ($0["name"] ?? $0["model"] ?? $0["id"]) as? String }
        let timings = timingsJSON as? [String: Any] ?? object["timings"] as? [String: Any] ?? [:]
        let promptMS = number(timings["prompt_ms"] ?? timings["prompt_eval_duration"])
        let predictedMS = number(timings["predicted_ms"] ?? timings["eval_duration"])
        let predictedN = number(timings["predicted_n"] ?? timings["eval_count"])
        let metricValues = parsePrometheus(metricsText)
        let contextUsed = integer(object["n_past"] ?? slots.compactMap { integer($0["n_past"]) }.max())
        let contextLimit = integer(object["n_ctx"] ?? object["context_length"] ?? slots.compactMap { integer($0["n_ctx"]) }.max())
        let queue = integer(object["queue"] ?? object["queue_depth"])
            ?? metricValues.first(suffix: "requests_waiting").map(Int.init)
        let tokensPerSecond = number(timings["predicted_per_second"] ?? timings["tokens_per_second"])
            ?? ((predictedMS ?? 0) > 0 && predictedN != nil ? predictedN! / predictedMS! * 1_000 : nil)
        return LocalRuntimeSnapshot(
            health: object.isEmpty && slots.isEmpty ? nil : .ready,
            loadedModelIDs: Array(Set(loaded)).sorted(),
            ttftMilliseconds: promptMS,
            tokensPerSecond: tokensPerSecond,
            contextUsed: contextUsed,
            contextLimit: contextLimit,
            queueDepth: queue,
            cpuPercent: number(object["cpu_percent"]) ?? metricValues.first(suffix: "cpu_percent"),
            gpuPercent: number(object["gpu_percent"]) ?? metricValues.first(suffix: "gpu_percent")
        )
    }

    static func tokenCount(_ json: Any?) -> Int? {
        if let object = json as? [String: Any] {
            if let tokens = object["tokens"] as? [Any] { return tokens.count }
            return integer(object["count"] ?? object["n_tokens"])
        }
        if let tokens = json as? [Any] { return tokens.count }
        return nil
    }

    private static func number(_ value: Any?) -> Double? {
        if let value = value as? NSNumber { return value.doubleValue.isFinite ? value.doubleValue : nil }
        if let value = value as? String, let parsed = Double(value), parsed.isFinite { return parsed }
        return nil
    }

    private static func integer(_ value: Any?) -> Int? { number(value).map(Int.init) }

    private static func parsePrometheus(_ text: String?) -> [String: Double] {
        guard let text else { return [:] }
        return text.split(separator: "\n").reduce(into: [:]) { result, line in
            guard !line.hasPrefix("#") else { return }
            let parts = line.split(whereSeparator: \.isWhitespace)
            guard parts.count >= 2, let value = Double(parts.last!) else { return }
            result[String(parts[0])] = value
        }
    }
}

private extension Dictionary where Key == String, Value == Double {
    func first(suffix: String) -> Double? { first { $0.key.hasSuffix(suffix) }?.value }
}

enum LocalEngineRuntimeClient {
    static func status(
        endpoint: URL,
        engine: LocalEngineKind,
        requested: RelayRequestedConfig,
        apiKey: String = "",
        session: URLSession = .shared
    ) async -> LocalRuntimeSnapshot {
        let statusPath: String
        switch engine {
        case .llamacpp: statusPath = "/slots"
        case .ollama: statusPath = "/api/ps"
        case .lmstudio: statusPath = "/api/v1/models"
        case .vllm: statusPath = "/v1/models"
        case .openwebui: statusPath = "/api/models"
        }
        let statusJSON = try? await get(
            endpoint: endpoint, path: statusPath, requested: requested,
            apiKey: apiKey, session: session
        ).json
        let metrics = engine == .llamacpp || engine == .vllm
            ? try? await get(
                endpoint: endpoint, path: "/metrics", requested: requested,
                apiKey: apiKey, session: session
            ).text
            : nil
        return LocalRuntimeParser.snapshot(engine: engine, statusJSON: statusJSON, metricsText: metrics)
    }

    static func preflight(
        endpoint: URL,
        engine: LocalEngineKind,
        prompt: String,
        contextLimit: Int?,
        requested: RelayRequestedConfig,
        session: URLSession = .shared
    ) async -> LocalPromptPreflight {
        guard engine == .llamacpp else { return .unavailable }
        do {
            let templated = try await postJSON(
                endpoint: endpoint,
                path: "/apply-template",
                body: ["prompt": prompt, "add_generation_prompt": true],
                requested: requested,
                session: session
            )
            let finalPrompt = (templated as? [String: Any])?["prompt"] as? String ?? prompt
            let tokenized = try await postJSON(
                endpoint: endpoint,
                path: "/tokenize",
                body: ["content": finalPrompt],
                requested: requested,
                session: session
            )
            guard let count = LocalRuntimeParser.tokenCount(tokenized) else { return .unavailable }
            let observedLimit: Int?
            if let contextLimit {
                observedLimit = contextLimit
            } else {
                observedLimit = await status(
                    endpoint: endpoint, engine: engine, requested: requested, session: session
                ).contextLimit
            }
            return .supported(tokens: count, contextLimit: observedLimit, exceedsContext: observedLimit.map { count > $0 } ?? false)
        } catch {
            return .unavailable
        }
    }

    static func promptCache(
        endpoint: URL,
        slotID: Int,
        action: String,
        cacheName: String,
        requested: RelayRequestedConfig,
        session: URLSession = .shared
    ) async throws {
        guard action == "save" || action == "restore", slotID >= 0 else { throw URLError(.badURL) }
        let encoded = cacheName.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? ""
        _ = try await postJSON(
            endpoint: endpoint,
            path: "/slots/\(slotID)?action=\(action)&filename=\(encoded)",
            body: [:],
            requested: requested,
            session: session
        )
    }

    private static func postJSON(
        endpoint: URL,
        path: String,
        body: [String: Any],
        requested: RelayRequestedConfig,
        session: URLSession
    ) async throws -> Any? {
        guard let url = URL(string: path, relativeTo: endpoint)?.absoluteURL else { throw URLError(.badURL) }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 5
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.applyRelaySecurityMode(requested)
        let (data, response) = try await session.relayData(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else { throw URLError(.badServerResponse) }
        return try? JSONSerialization.jsonObject(with: data)
    }

    private static func get(
        endpoint: URL,
        path: String,
        requested: RelayRequestedConfig,
        apiKey: String = "",
        session: URLSession
    ) async throws -> (json: Any?, text: String) {
        guard let url = URL(string: path, relativeTo: endpoint)?.absoluteURL else { throw URLError(.badURL) }
        var request = URLRequest(url: url)
        request.timeoutInterval = 5
        if !apiKey.isEmpty { request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization") }
        request.applyRelaySecurityMode(requested)
        let (data, response) = try await session.relayData(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else { throw URLError(.badServerResponse) }
        return (try? JSONSerialization.jsonObject(with: data), String(decoding: data, as: UTF8.self))
    }
}

File: android/app/src/main/java/ai/oriveo/community/core/provider/LocalEngineRuntime.kt
package ai.oriveo.community.core.provider

import android.content.Context
import io.ktor.client.HttpClient
import io.ktor.client.request.post
import io.ktor.client.request.get
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.client.request.header
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

@Serializable
data class LocalRuntimeSettings(
    val contextLength: Int? = null,
    val keepAlive: String? = null,
    val speculativeModel: String? = null,
    val cacheReuseTokens: Int? = null,
)

class LocalRuntimeSettingsStore private constructor(private val context: Context, private val json: Json) {
    private val preferences = context.applicationContext.getSharedPreferences("local_runtime_settings", Context.MODE_PRIVATE)

    fun load(endpointFingerprint: String): LocalRuntimeSettings = preferences.getString(endpointFingerprint, null)?.let {
        runCatching { json.decodeFromString(LocalRuntimeSettings.serializer(), it) }.getOrNull()
    } ?: LocalRuntimeSettings()

    fun save(endpointFingerprint: String, settings: LocalRuntimeSettings) {
        preferences.edit().putString(endpointFingerprint, json.encodeToString(LocalRuntimeSettings.serializer(), settings)).apply()
    }

    fun remove(endpointFingerprint: String) { preferences.edit().remove(endpointFingerprint).apply() }

    companion object {
        fun from(context: Context, json: Json = Json { ignoreUnknownKeys = true }) = LocalRuntimeSettingsStore(context, json)
        fun fingerprint(endpoint: String, engine: LocalEngineKind): String =
            (engine.name + "|" + endpoint).fold(0xcbf29ce484222325UL) { hash, char -> (hash xor char.code.toULong()) * 0x100000001b3UL }.toString(16)
    }
}

data class LocalRuntimeSnapshot(
    val health: LocalEngineState? = null,
    val loadedModelIDs: List<String> = emptyList(),
    val ttftMilliseconds: Double? = null,
    val tokensPerSecond: Double? = null,
    val contextUsed: Int? = null,
    val contextLimit: Int? = null,
    val queueDepth: Int? = null,
    val cpuPercent: Double? = null,
    val gpuPercent: Double? = null,
)

sealed interface LocalPromptPreflight {
    data class Supported(val tokens: Int, val contextLimit: Int?, val exceedsContext: Boolean) : LocalPromptPreflight
    data object Unavailable : LocalPromptPreflight
}

/** Parses real engine responses; missing measurements remain null and are rendered as unavailable. */
object LocalRuntimeParser {
    fun snapshot(status: JsonElement?, metricsText: String? = null, timings: JsonObject? = null): LocalRuntimeSnapshot {
        val root = status as? JsonObject ?: JsonObject(emptyMap())
        val slots = (root["slots"] as? JsonArray) ?: (status as? JsonArray) ?: JsonArray(emptyList())
        val models = (root["models"] as? JsonArray) ?: (root["data"] as? JsonArray) ?: JsonArray(emptyList())
        val loaded = models.mapNotNull { row ->
            val objectValue = row as? JsonObject ?: return@mapNotNull null
            listOf("name", "model", "id").firstNotNullOfOrNull { objectValue[it]?.jsonPrimitive?.content }
        }.distinct().sorted()
        val timing = timings ?: (root["timings"] as? JsonObject) ?: JsonObject(emptyMap())
        val predictedMilliseconds = timing.number("predicted_ms") ?: timing.durationMilliseconds("eval_duration")
        val predictedTokens = timing.number("predicted_n") ?: timing.number("eval_count")
        val metrics = parsePrometheus(metricsText)
        val contextUsed = root.integer("n_past") ?: slots.mapNotNull { (it as? JsonObject)?.integer("n_past") }.maxOrNull()
        val contextLimit = root.integer("n_ctx") ?: root.integer("context_length")
            ?: slots.mapNotNull { (it as? JsonObject)?.integer("n_ctx") }.maxOrNull()
        return LocalRuntimeSnapshot(
            health = if (root.isEmpty() && slots.isEmpty()) null else LocalEngineState.Ready,
            loadedModelIDs = loaded,
            ttftMilliseconds = timing.number("prompt_ms") ?: timing.durationMilliseconds("prompt_eval_duration"),
            tokensPerSecond = timing.number("predicted_per_second") ?: timing.number("tokens_per_second")
                ?: if (predictedMilliseconds != null && predictedMilliseconds > 0 && predictedTokens != null) predictedTokens / predictedMilliseconds * 1_000 else null,
            contextUsed = contextUsed,
            contextLimit = contextLimit,
            queueDepth = root.integer("queue") ?: root.integer("queue_depth") ?: metrics.firstSuffix("requests_waiting")?.toInt(),
            cpuPercent = root.number("cpu_percent") ?: metrics.firstSuffix("cpu_percent"),
            gpuPercent = root.number("gpu_percent") ?: metrics.firstSuffix("gpu_percent"),
        )
    }

    fun tokenCount(value: JsonElement?): Int? = when (value) {
        is JsonArray -> value.size
        is JsonObject -> (value["tokens"] as? JsonArray)?.size
            ?: value["count"]?.jsonPrimitive?.intOrNull
            ?: value["n_tokens"]?.jsonPrimitive?.intOrNull
        else -> null
    }

    private fun JsonObject.number(key: String): Double? = (this[key] as? JsonPrimitive)?.doubleOrNull
    private fun JsonObject.integer(key: String): Int? = (this[key] as? JsonPrimitive)?.intOrNull
    private fun JsonObject.durationMilliseconds(key: String): Double? = number(key)?.let { if (it > 1_000_000) it / 1_000_000 else it }
    private fun parsePrometheus(text: String?): Map<String, Double> = text.orEmpty().lineSequence().mapNotNull { line ->
        if (line.startsWith('#')) return@mapNotNull null
        val parts = line.trim().split(Regex("\\s+"))
        if (parts.size < 2) null else parts.first() to (parts.last().toDoubleOrNull() ?: return@mapNotNull null)
    }.toMap()
    private fun Map<String, Double>.firstSuffix(suffix: String): Double? = entries.firstOrNull { it.key.endsWith(suffix) }?.value
}

class LocalEngineRuntimeClient(private val client: HttpClient, private val json: Json) {
    suspend fun status(endpoint: String, engine: LocalEngineKind, apiKey: String = ""): LocalRuntimeSnapshot {
        val path = when (engine) {
            LocalEngineKind.LlamaCpp -> "/slots"
            LocalEngineKind.Ollama -> "/api/ps"
            LocalEngineKind.LmStudio -> "/api/v1/models"
            LocalEngineKind.Vllm -> "/v1/models"
            LocalEngineKind.OpenWebUI -> "/api/models"
        }
        val status = getBody(endpoint, path, apiKey)?.let { runCatching { json.parseToJsonElement(it) }.getOrNull() }
        val metrics = if (engine == LocalEngineKind.LlamaCpp || engine == LocalEngineKind.Vllm) getBody(endpoint, "/metrics", apiKey) else null
        return LocalRuntimeParser.snapshot(status, metrics)
    }

    suspend fun preflight(endpoint: String, engine: LocalEngineKind, prompt: String, contextLimit: Int?): LocalPromptPreflight {
        if (engine != LocalEngineKind.LlamaCpp) return LocalPromptPreflight.Unavailable
        return runCatching {
            val templated = postJSON(endpoint, "/apply-template", buildJsonObject {
                put("prompt", prompt)
                put("add_generation_prompt", true)
            })
            val finalPrompt = (templated as? JsonObject)?.get("prompt")?.jsonPrimitive?.content ?: prompt
            val tokenized = postJSON(endpoint, "/tokenize", buildJsonObject { put("content", finalPrompt) })
            val count = LocalRuntimeParser.tokenCount(tokenized) ?: return@runCatching LocalPromptPreflight.Unavailable
            val observedLimit = contextLimit ?: status(endpoint, engine).contextLimit
            LocalPromptPreflight.Supported(count, observedLimit, observedLimit?.let { count > it } ?: false)
        }.getOrDefault(LocalPromptPreflight.Unavailable)
    }

    suspend fun promptCache(endpoint: String, slotID: Int, action: String, cacheName: String) {
        require(slotID >= 0 && action in setOf("save", "restore"))
        postJSON(endpoint, "/slots/$slotID?action=$action&filename=${java.net.URLEncoder.encode(cacheName, Charsets.UTF_8.name())}", JsonObject(emptyMap()))
    }

    private suspend fun postJSON(endpoint: String, path: String, body: JsonObject): JsonElement? {
        val response = client.post(endpoint.trimEnd('/') + path) {
            header(HttpHeaders.ContentType, ContentType.Application.Json)
            setBody(body.toString())
        }
        if (response.status.value !in 200..299) error("local runtime request rejected")
        return response.bodyAsText().takeIf(String::isNotBlank)?.let(json::parseToJsonElement)
    }

    private suspend fun getBody(endpoint: String, path: String, apiKey: String = ""): String? = runCatching {
        val response = client.get(endpoint.trimEnd('/') + path) {
            if (apiKey.isNotEmpty()) header(HttpHeaders.Authorization, "Bearer $apiKey")
        }
        if (response.status.value !in 200..299) return@runCatching null
        response.bodyAsText()
    }.getOrNull()
}
